### ISSUE TITLE
Fix race condition on alert_manager::maybe_notify()

### ISSUE THREAD

**Issue by fernando-rodriguez** [arvidn/libtorrent#2934]
**Please provide the following information**

libtorrent version (or branch): RC_1_1

platform/architecture: All

compiler and compiler version: All

There is a race condition on `alert_manager::maybe_notify()` caused by unlocking the mutex before calling `plugin::on_alert()`. If another thread pops alerts twice the alert allocations can become invalid before `on_alert()` completes. It can also result in `on_alert()` being called after the bit for that alert has been cleared.

This cannot normally happen because all alerts are posted from the same thread, but there is a deprecated alert being posted from the IO thread.

**Comment by fernando-rodriguez** [comment]
It would also affect the `set_alert_notify()` callback.

**Comment by fernando-rodriguez** [comment]
I think we can use a separate recursive mutex just to protect the allocator.

**Comment by arvidn** [comment]
I’m in the camp of thinking that recursive mutexes are a symptom of a design issue. Thanks for reporting this, my intention and hope is that there will be more threads in the future, so I’ll try to fix this

**Comment by arvidn** [comment]
there is a fundamental problem here. If the mutex is held while calling the callback, the callback can't do anything that may post another alert (since it's already locked, it would deadlock). That's clearly not a good situation since it's quite likely the callback would want to do something interesting there.

On the other hand, if the mutex is unlocked, not only can another thread cause the passed in alert to become dangling, having a sufficient number of alerts be posted will also cause the queue to grow and possibly reallocate the alert, also making it a dangling pointer.

**Comment by fernando-rodriguez** [comment]
Actually strike the first part. In this case I think it's safe because it's never called with the mutex locked.

**Comment by arvidn** [comment]
I hadn't considered that a recursive mutex may have odd interactions with condition variables. But, as you say, as long as the thread calling `wait()` has only locked the mutex once, it should be safe.

As for the second part, the future I envision would decouple peer connections and not have them block on each other, so I don't think it would be important to be re-entrant from different threads. I consider it a hack to need a recursive_mutex at all, that comes out of having arbitrary user code run in-thread. Normally this would be avoided by posting messages and deferring the handler (how alerts are meant to be handled).

**Comment by fernando-rodriguez** [comment]
This is not part of this report or your PR but I think there's another race condition on `condition_var::wait()` and `wait_for()` on Windows and Beos because the mutex is unlocked before waiting for the semaphore, no?

> As for the second part, the future I envision would decouple peer connections and not have them block on each other, so I don't think it would be important to be re-entrant from different threads.

But they will be coupled by the alert manager, no? The only sure chance a peer thread will have of acquiring the mutex will be when the other peer threads are waiting for data. So a peer that does not have to wait long for data will have a very high chance of getting preempted while holding the lock so all other peers loose their turns, even on other cores. By the time the scheduler goes around the peer will likely have more data ready so the process can repeat many times. It will be much worse with realtime scheduling but still a problem everywhere. To minimize the chance of this happening the critical sections need to be very small or

**Comment by arvidn** [comment]
that's not my understanding. since a semaphore is a counter, different from a signal, it doesn't matter whether the signalling thread increments the counter first or whether the waiting thread attempts to decrement it first. If the signalling thread gets there first, it will just increment the counter as many times as there are waiters, and once the waiters get there, they can decrement the semaphore without waiting. If the waiters get there first, they will all try to decrement a semaphore that's zero, and they will block until the signalling thread increments the semaphore.

My understanding of the significance of the atomicity of unlocking the mutex and to wait for the signal, is because the signal is missed if not waited for when it happens. This is not the case with semaphores.

**Comment by arvidn** [comment]
in my measurements, the work being done under the alert managers mutex is negligible (at least after the optimisations of allocations of alerts and their data) compared to the socket send/receive (memcopy to/from the kernel), the disk I/O and the SHA-1 hashing. Granted, it's mostly the first one that happens in the network thread, where the alerts are posted from.

**Comment by fernando-rodriguez** [comment]
What if `notify()` or `notify_all()` is called twice before the waiter gets there? Or if it's called once while the other thread is incrementing or decrementing the counter?

**Comment by fernando-rodriguez** [comment]
> Granted, it's mostly the first one that happens in the network thread, where the alerts are posted from.

That can be very fast. Some x86 chips can move 512-bit in one instruction and that data may already be on an L3 cache, while some strings have to be copied 1 byte at the time. But even if it was 1% you have a problem with the mutex. A time slice is 1 to a few ms depending on the system. With that many threads, what are the chances that one thread will get to stay that long on-cpu without colliding with the others? I'm not sure how to calculate it but it's slim. Besides that each thread has a 1/100 chance of getting preempted while the mutex is locked. My dumb math says that if the critical section was 1%, with 100 threads you'll have 100% probability of having only 1 peer thread on-cpu at any given time. I think it's actually higher but I'm not sure.

With the lock-free patch you'll be able to have 1 thread on each available core at most times, except when the user calls a function that takes an exclusive lock, such as `set_notify_function()`, `set_dispatch_function()`, and `set_queue_size_limit()`. When there are collissions on the small critical section we wait on a spinlock. The possibility of the thread that the spinlock is waiting on getting interrupted while we wait is slim, so most of the time each peer will get a whole timeslice. If it does happens it results in the spinlock yielding the cpu (so it becomes a mutex) but other threads can still make progress up to the spinlock and after 1 scheduler cycle all threads can make progress again. As it is, even with the lock-free patch you may have problems with that many threads, because the spinlock where `m_write_slot` is incremented will need to be fair. Also you may not want to use that many threads, you'll only get benefit of 1 thread per cpu so what you probably want is one thread per cpu running co-routines for each peer. Boost has a co-routine implementation.

**Comment by arvidn** [comment]
this race was fixed by the recursive mutex
